File: ros2_ws/src/perception_pkg/perception_pkg/depth_to_pointcloud_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from sensor_msgs.msg import Image, CameraInfo, PointCloud2
from image_geometry import PinholeCameraModel
import cv2
from cv_bridge import CvBridge
import numpy as np
import sensor_msgs_py.point_cloud2 as pc2

class DepthToPointCloudNode(Node):
    def __init__(self):
        super().__init__('depth_to_pointcloud_node')

        # Parameters matching your simulation's topic list
        self.declare_parameter('depth_topic', '/realsense/depth/image')
        self.declare_parameter('camera_info_topic', '/realsense/depth/camera_info')
        self.declare_parameter('pointcloud_topic', '/points')

        depth_topic = self.get_parameter('depth_topic').value
        camera_info_topic = self.get_parameter('camera_info_topic').value
        pointcloud_topic = self.get_parameter('pointcloud_topic').value

        # Tools
        self.bridge = CvBridge()
        self.camera_model = PinholeCameraModel()
        self.camera_info_received = False

        # Subscribers
        self.create_subscription(CameraInfo, camera_info_topic, self.camera_info_callback, 10)
        self.create_subscription(Image, depth_topic, self.depth_callback, 10)

        # Publisher
        self.pc_pub = self.create_publisher(PointCloud2, pointcloud_topic, 10)

    def camera_info_callback(self, msg: CameraInfo):
        if not self.camera_info_received:
            self.camera_model.fromCameraInfo(msg)
            self.camera_info_received = True
            self.get_logger().info('Camera info received and camera model set.')

    def depth_callback(self, msg: Image):
        if not self.camera_info_received:
            return

        # 1. Convert ROS Image to OpenCV/Numpy array
        try:
            depth_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='passthrough')
        except Exception as e:
            self.get_logger().error(f"CvBridge Error: {e}")
            return

        # Ensure we are doing math on floats
        depth_image = depth_image.astype(np.float32)

        # ==========================================
        # SCALING FIX:
        # If your debug prints show max values > 100, your simulation is outputting millimeters.
        # UNCOMMENT THE LINE BELOW TO CONVERT TO METERS:
        depth_image = depth_image / 1000.0
        # ==========================================

        # 2. Vectorized Math Setup
        height, width = depth_image.shape
        u, v = np.meshgrid(np.arange(width), np.arange(height))
        
        z = depth_image.flatten()
        u = u.flatten()
        v = v.flatten()

        # 3. Debug Print (Prints roughly once every 30 frames)
        if np.random.rand() < 0.03:
            max_z = np.nanmax(z) if len(z) > 0 else 0.0
            min_z = np.nanmin(z) if len(z) > 0 else 0.0
            self.get_logger().info(f"RAW DEPTH -> Min: {min_z:.2f}, Max: {max_z:.2f}")

        # 4. Strict Filtering (Remove NaNs, Inf, and Out-of-Bounds values)
        # Assuming values are in meters: keep points between 0.1m and 20.0m
        valid_mask = (z > 0.1) & (z < 20.0) & np.isfinite(z)
        
        z_valid = z[valid_mask]
        u_valid = u[valid_mask]
        v_valid = v[valid_mask]

        # Early exit if all points were destroyed by the filter
        if len(z_valid) == 0:
            return

        # 5. Apply Pinhole Camera Model Math (Optical Frame)
        x_opt = (u_valid - self.camera_model.cx()) * z_valid / self.camera_model.fx()
        y_opt = (v_valid - self.camera_model.cy()) * z_valid / self.camera_model.fy()
        z_opt = z_valid
        
        # 6. Convert to ROS Standard Frame (X-forward, Y-left, Z-up)
        # We rotate the optical coordinates to match the drone's body frame
        x_body = z_opt
        y_body = -x_opt
        z_body = -y_opt
        
        # Stack into list of [x, y, z] arrays
        points = np.vstack((x_body, y_body, z_body)).T.tolist()

        # 6. Override the frame_id to match the simulation's actual TF Tree
        msg.header.frame_id = 'Quadrotor/DepthCamera'
        
        # 7. Create and Publish the PointCloud2 message
        pc2_msg = pc2.create_cloud_xyz32(msg.header, points)
        self.pc_pub.publish(pc2_msg)

def main(args=None):
    rclpy.init(args=args)
    node = DepthToPointCloudNode()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws/src/perception_cpp_pkg/src/depth_to_pointcloud_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/point_cloud2_iterator.hpp>
#include <cv_bridge/cv_bridge.hpp>
#include <image_geometry/pinhole_camera_model.hpp>

class DepthToPointCloudNode : public rclcpp::Node {
public:
    DepthToPointCloudNode() : Node("depth_to_pointcloud_node"), camera_info_received_(false) {
        // Declare and get parameters
        this->declare_parameter<std::string>("depth_topic", "/realsense/depth/image");
        this->declare_parameter<std::string>("camera_info_topic", "/realsense/depth/camera_info");
        this->declare_parameter<std::string>("pointcloud_topic", "/points");

        std::string depth_topic = this->get_parameter("depth_topic").as_string();
        std::string camera_info_topic = this->get_parameter("camera_info_topic").as_string();
        std::string pointcloud_topic = this->get_parameter("pointcloud_topic").as_string();

        rclcpp::QoS qos(10);

        // Subscribers & Publishers
        cam_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
            camera_info_topic, qos, std::bind(&DepthToPointCloudNode::cameraInfoCallback, this, std::placeholders::_1));
        
        depth_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            depth_topic, qos, std::bind(&DepthToPointCloudNode::depthCallback, this, std::placeholders::_1));

        pc_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(pointcloud_topic, qos);
        
        RCLCPP_INFO(this->get_logger(), "C++ Depth to PointCloud Node Started.");
    }

private:
    void cameraInfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg) {
        if (!camera_info_received_) {
            camera_model_.fromCameraInfo(*msg);
            camera_info_received_ = true;
            RCLCPP_INFO(this->get_logger(), "Camera info received and model set.");
        }
    }

    void depthCallback(const sensor_msgs::msg::Image::SharedPtr msg) {
        if (!camera_info_received_) return;

        cv_bridge::CvImagePtr cv_ptr;
        try {
            // Safely convert any depth format into a float matrix
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::TYPE_32FC1); 
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        cv::Mat depth_image = cv_ptr->image;

        // Initialize PointCloud2 message
        sensor_msgs::msg::PointCloud2 cloud_msg;
        cloud_msg.header = msg->header;
        cloud_msg.header.frame_id = "Quadrotor/DepthCamera"; // Override TF frame
        cloud_msg.height = 1;
        cloud_msg.width = 0; // Will be updated
        cloud_msg.is_dense = false;
        cloud_msg.is_bigendian = false;

        // Efficiently allocate memory for the points
        sensor_msgs::PointCloud2Modifier modifier(cloud_msg);
        modifier.setPointCloud2FieldsByString(1, "xyz");
        modifier.resize(depth_image.rows * depth_image.cols);

        sensor_msgs::PointCloud2Iterator<float> iter_x(cloud_msg, "x");
        sensor_msgs::PointCloud2Iterator<float> iter_y(cloud_msg, "y");
        sensor_msgs::PointCloud2Iterator<float> iter_z(cloud_msg, "z");

        size_t valid_points = 0;

        // Iterate through all pixels
        for (int v = 0; v < depth_image.rows; ++v) {
            for (int u = 0; u < depth_image.cols; ++u) {
                float z = depth_image.at<float>(v, u);
                
                // Millimeter to Meter conversion (if values are huge like we saw in Python)
                if (z > 100.0) { 
                    z = z / 1000.0;
                }

                // Strict Filter: Remove NaN, Inf, and limits
                if (std::isfinite(z) && z > 0.1 && z < 50.0) {
                    
                    // Pinhole math (Optical Frame)
                    float x_opt = (u - camera_model_.cx()) * z / camera_model_.fx();
                    float y_opt = (v - camera_model_.cy()) * z / camera_model_.fy();
                    
                    // Rotate to ROS Body Frame (X-forward, Y-left, Z-up)
                    *iter_x = z;
                    *iter_y = -x_opt;
                    *iter_z = -y_opt;

                    ++iter_x; ++iter_y; ++iter_z;
                    ++valid_points;
                }
            }
        }

        if (valid_points == 0) return;

        // Shrink cloud to actual number of valid points to save bandwidth
        modifier.resize(valid_points); 
        pc_pub_->publish(cloud_msg);
    }

    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr cam_info_sub_;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr depth_sub_;
    rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr pc_pub_;
    
    image_geometry::PinholeCameraModel camera_model_;
    bool camera_info_received_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DepthToPointCloudNode>());
    rclcpp::shutdown();
    return 0;
}